numpy/random/mtrand/mtrand.pyx

```diff
@@ -5064,7 +5064,8 @@ cdef class RandomState:
             x_ptr = <char*><size_t>x.ctypes.data
             stride = x.strides[0]
             itemsize = x.dtype.itemsize
-            buf = np.empty_like(x[0])  # GC'd at function exit
+            # Ensure that buf has the same dtype as x
+            buf = np.empty_like(x[:1])  # GC'd at function exit
```

Comment:
Ah, of course, the view will incref its elements, then they get shuffled under its feet and the decref goes to the wrong one. Easiest solution is likely to just avoid that 1D optimization for object arrays by checking `arr.dtype.hasobject` first.